adoptingerlang.org

```diff
@@ -1331,6 +1331,364 @@ Repeat this exercise from time to time, and eventually do some [[https://princip
+
+**** Proxy Support
+
+Many workplaces enforce very strict firewall rules, to the point where all incoming and outgoing data must be intercepted and monitored. Generally, these places will not be totally isolated from the public Internet, but will require the use of proxy servers to make all connections work.
+
+It is rather standard for open source program to respect both the =HTTP_PROXY= and =HTTPS_PROXY= environment variable. When those are set in the development environment you use, rebar3 will make sure that all the communications it makes talking to the outside world will reuse these proxies.
+
+This should let you properly work in line with your IT department's policies. In some cases, that won't even be enough.
+
+**** Private Hex Mirrors
+
+Some corporations go a step further and segment their internal network from the public Internet. All data that comes on site has to be inspected and hosted independently, without a chance to talk to a common code repository like github, gitlab, or hex. Another interesting case is build servers, where you might want to prevent all connectivity to the outside world both for safety and for repeatability reasons.
+
+For such setups, two approaches tend to be used: vendoring in a monorepo, which will be covered in the next section, and through privately hosted package indexes, which we'll cover here.
+
+The idea behind a privately hosted index is that all packages and dependencies to be used in a project need to be fully vetted. You might want to give it a technical review for code quality, a security assessment, or have corporate lawyers look at code for licensing or patent issues. Then, only the packages of the right version can be used. This kind of index is frequently accepted for hermetic builds with the assumption that it either runs locally on each build server, or within a private network that is as tightly monitored as build servers.
+
+Rebar3 supports this use case. If you want to enable it, you will need to first set up a private hex instance.
+TODO: fill instructions from what Starbelly had written.
+TODO: fill instructions to use minirepo
+
+**** About Monorepos
+
+If you work in a corporate environment with a monorepo, rebar3 is not the best tool for the job. Mostly, this comes down to the fact that most companies using a monorepo have a lot of custom tooling with very custom workflows, large codebases, and a very strong propension to never share access with maintainers of rebar3. Until it becomes possible for maintainers to get access, little will be possible to do on that front.
+
+Various commercial users who rely on rebar3 despite using monorepos have gotten successful builds with a combination of using =_checkout= dependencies, along with modifying the =_build= directory (TODO: add option)
```

Comment:
We don't want to encourage this though, right?

I also think we should make clear an umbrella project is not the same thing as a monorepo.

I have a lot of questions about why rebar3 doesn't work for these monorepos too.. maybe they aren't using releases, I guess. Having a rebar3  project in a monorepo that builds a release for deployment shouldn't be an issue, aside from there not being a Bazel/Buck plugin.